substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -209,83 +215,110 @@ where
 			chain_head_follow.generate_events(sink, sub_data).await;
 
 			subscriptions.remove_subscription(&sub_id);
+			rpc_connections.unregister_token(connection_id, &sub_id);
 			debug!(target: LOG_TARGET, "[follow][id={:?}] Subscription removed", sub_id);
 		};
 
 		self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
 	}
 
-	fn chain_head_unstable_body(
+	async fn chain_head_unstable_body(
 		&self,
+		connection_details: ConnectionDetails,
 		follow_subscription: String,
 		hash: Block::Hash,
 	) -> ResponsePayload<'static, MethodResponse> {
-		let mut block_guard = match self.subscriptions.lock_block(&follow_subscription, hash, 1) {
-			Ok(block) => block,
-			Err(SubscriptionManagementError::SubscriptionAbsent) |
-			Err(SubscriptionManagementError::ExceededLimits) =>
-				return ResponsePayload::success(MethodResponse::LimitReached),
-			Err(SubscriptionManagementError::BlockHashAbsent) => {
-				// Block is not part of the subscription.
-				return ResponsePayload::error(ChainHeadRpcError::InvalidBlock);
-			},
-			Err(_) => return ResponsePayload::error(ChainHeadRpcError::InvalidBlock),
-		};
+		if !self
+			.rpc_connections
+			.contains_token(connection_details.id(), &follow_subscription)
+		{
+			return ResponsePayload::success(MethodResponse::LimitReached);
```

Comment:
IS this the right error to return? If this `if` is true then it means a connection is trying to call `chainHead_body` but hasn't called `chainHead_follo` yet I think?

Reply:
Yep that makes sense! 

When this check is true, the current connection hasn't obtained the `follow_subscription` ID by making a call to the `chainHead_follow` first. This could mean that the user is guessing the follow subscription of another user, or that the follow subscription was leaked from another valid subscription.


> (I also can't offhand see where we do anything to limit the number of chainHead_follow subscriptions, which I assumed might exist for us to emit this error, but maybe I'm either missing something or it's not present yet :))

I'll add it in this PR since it shouldn't be that complicated, initially wanted to do a followup for this :D

Reply:
One thing I didn't realise when I saw this first is that the spec actually says to return `limitReached` when the followSubscription is invalid/stale, so this all maekssense to me now (a nit here might help make clear why we return an error that feels unrelaetd at a glance eg "the spec says to return LimitReached if followSubscription is invalid or stale" :))